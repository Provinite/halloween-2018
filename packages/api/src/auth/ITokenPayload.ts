/**
 * Data stored in JWT payload.
 */
export interface ITokenPayload {
  /** User's deviantart api access token */
  accessToken: string;
  /** Subject, user's uuid */
  sub: string;
  /** Date (seconds since epoch) indicating issuance time. */
  iat: number;
  /** Date (seconds since epoch) indicating expiry time */
  exp: number;
}
