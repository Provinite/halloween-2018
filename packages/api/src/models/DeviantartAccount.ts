import {
  Entity,
  OneToOne,
  RelationId,
  PrimaryColumn,
  JoinColumn
} from "typeorm";
import { User } from "./User";

@Entity()
export class DeviantartAccount {
  /**
   * The user who owns this DA association
   */
  @OneToOne(type => User, {
    nullable: false
  })
  @JoinColumn()
  user!: User;

  /**
   * The ID of the user for this account
   */
  @RelationId((deviantartAccount: DeviantartAccount) => deviantartAccount.user)
  userId!: number;

  /**
   * The deviantart UUID for this account
   */
  @PrimaryColumn()
  deviantartUuid!: string;
}
