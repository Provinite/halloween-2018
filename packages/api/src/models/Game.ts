import { Column, Entity, PrimaryGeneratedColumn } from "typeorm";

/**
 * A game represents a single event or giveaway and provides a containing scope
 * for prize pools and draws. Games provide a way to configure win rates and
 * draw schedules as well as run multiple concurrent giveaways.
 */
@Entity()
export class Game {
  @PrimaryGeneratedColumn()
  id!: number;

  /** The name of the game. */
  @Column({ nullable: false, unique: true })
  name!: string;

  /** Description for the game. */
  @Column({ nullable: false })
  description!: string;

  /** Contact details for the game's owner */
  @Column({ nullable: false })
  contact!: string;

  /** Datetime at which the game opens. */
  @Column({ type: "time without time zone", nullable: true })
  startDate: Date | null = null;

  /** Datetime at which the game closes. */
  @Column({ type: "time without time zone", nullable: true })
  endDate: Date | null = null;

  /**
   * Crontab style string representing the draw reset schedule.
   * Check out https://crontab.guru for details.
   * @example
   * 0 12 1,15 * * // noon on the first and fifteenth of every month
   * 0 12 * * * // every day at noon
   */
  @Column({
    default: "0 12 * * *"
  })
  drawResetSchedule!: string;

  /**
   * How many draws a player gets per reset.
   */
  @Column({ default: 1 })
  drawsPerReset!: number;

  /**
   * How many draws a VIP gets per reset.
   */
  @Column({ default: 2 })
  vipDrawsPerReset!: number;

  /**
   * The win rate (percentage) for this event.
   * @example
   * 0.75 // 75% winrate
   * 0.5 // 50% winrate
   * 0.3333 // 33.33% winrate
   * 1 // 100% winrate
   */
  @Column({
    default: 0.5,
    type: "double precision"
  })
  winRate!: number;
}
