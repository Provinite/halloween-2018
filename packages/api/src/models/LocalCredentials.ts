import {
  Entity,
  Column,
  OneToOne,
  RelationId,
  JoinColumn,
  PrimaryGeneratedColumn
} from "typeorm";
import { User } from "./User";

@Entity()
/**
 * Login credentials for users that use local authentication.
 */
export class LocalCredentials {
  /**
   * The user that is authenticated by these credentials.
   */
  @OneToOne(type => User, {
    nullable: false
  })
  @JoinColumn()
  user!: User;

  @RelationId((localCredentials: LocalCredentials) => localCredentials.user)
  userId!: number;

  @PrimaryGeneratedColumn()
  id!: number;

  /**
   * The login (username/email/w/e for this user)
   */
  @Column({
    unique: true,
    length: 255
  })
  principal!: string;

  /**
   * The hashed password for this user.
   */
  @Column({ nullable: false })
  password!: string;
}
