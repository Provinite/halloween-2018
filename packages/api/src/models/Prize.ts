import { Column, Entity, PrimaryGeneratedColumn } from "typeorm";

@Entity()
export class Prize {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  name: string;

  @Column()
  description: string;

  @Column()
  initialStock: number;

  @Column()
  currentStock: number;
}
