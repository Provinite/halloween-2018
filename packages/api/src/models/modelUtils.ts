import {
  Connection,
  EntityManager,
  getMetadataArgsStorage,
  Repository
} from "typeorm";
import { IModelClass } from ".";

export function getRepositoryFor<T>(
  orm: EntityManager | Connection,
  modelClass: IModelClass<T>
): Repository<T> {
  const customRepository = getMetadataArgsStorage().entityRepositories.find(
    repository => repository.entity === modelClass
  );
  if (customRepository) {
    return orm.getCustomRepository(customRepository.target);
  }
  return orm.getRepository(modelClass);
}
