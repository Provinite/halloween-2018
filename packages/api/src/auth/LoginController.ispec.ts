import { createTestUser } from "../test/integration/authHelpers";
import { testApi } from "../test/integration/supertestWrapper";
import { verify } from "jsonwebtoken";
import { User } from "../models";
import { ITokenPayload } from "./ITokenPayload";

describe("POST /login", () => {
  let user: User;
  const password = "password";
  beforeAll(async () => {
    user = await createTestUser({ password });
  });
  const endpoint = "/login";
  it("returns a JWT on success", async () => {
    const body = {
      principal: user.displayName,
      password: "password"
    };
    const result = await testApi()
      .post(endpoint)
      .send(body)
      .expect(200);

    const token = result.body.token as string;

    const tokenPayload = verify(
      token,
      process.env.cch2018_token_secret!
    ) as ITokenPayload;

    // it should contain the expected user id
    expect(tokenPayload.sub).toEqual(user.id);

    // it should be valid for 55 minutes
    expect(tokenPayload.exp - tokenPayload.iat).toEqual(55 * 60);

    // comparison is done using a 1-minute window for expiration time to make
    // these tests less likely to report false failures
    // if this test starts to fail, look into injecting a mock time service
    // and using it everywhere we use dates.
    const nowInSeconds = new Date().getTime() / 1000;

    // it should expire in more than 54 minutes
    const lowerBound = nowInSeconds + 54 * 60;
    expect(tokenPayload.exp).toBeGreaterThan(lowerBound);

    // it should expire in less than 56 minutes
    const upperBound = nowInSeconds + 56 * 60;
    expect(tokenPayload.exp).toBeLessThan(upperBound);
  });

  it("returns a 400 with invalid credentials", async () => {
    const body = {
      principal: "username",
      password: "Pa$$word!"
    };
    await testApi()
      .post(endpoint)
      .send(body)
      .expect(400);
  });

  it("returns a 400 on unknown user", async () => {
    const body = {
      principal: "Dwight Schrute",
      password: "SomethingFunny"
    };
    await testApi()
      .post(endpoint)
      .send(body)
      .expect(400);
  });

  it("returns a 400 with no request body", async () => {
    await testApi()
      .post(endpoint)
      .send()
      .expect(400);
  });

  it("validates principal and password", async () => {
    let body: { principal?: any; password?: any } = {
      principal: "one"
    };
    let result = await testApi()
      .post(endpoint)
      .send(body)
      .expect(400);
    expect(result.body).toMatchSnapshot();

    body = {
      password: 2
    };

    result = await testApi()
      .post(endpoint)
      .send(body)
      .expect(400);
    expect(result.body).toMatchSnapshot();
  });
});
