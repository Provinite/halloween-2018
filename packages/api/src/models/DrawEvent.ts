import {
  CreateDateColumn,
  Entity,
  ManyToOne,
  PrimaryGeneratedColumn,
  RelationId
} from "typeorm";
import { Game } from "./Game";
import { Prize } from "./Prize";
import { User } from "./User";

@Entity()
export class DrawEvent {
  // see constructor
  readonly isWin: boolean;

  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne(type => User, { eager: true })
  user: User;

  @ManyToOne(type => Game)
  game: Game;

  @RelationId((drawEvent: DrawEvent) => drawEvent.game)
  gameId: number;

  @CreateDateColumn()
  createDate: Date;

  @ManyToOne(type => Prize, { nullable: true, eager: true })
  prize: Prize;
  constructor() {
    // Define isWin on the instance directly so that it appears as an instance
    // field for things like JSON generation
    Object.defineProperty(this, "isWin", {
      get: () => this.prize !== null,
      enumerable: true
    });
  }
}
