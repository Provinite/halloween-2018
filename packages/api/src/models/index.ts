import { Prize } from "./Prize";
export { Prize } from "./Prize";
export { Role } from "./Role";
import { DrawEvent } from "./DrawEvent";
import { Role } from "./Role";
import { User } from "./User";
export { DrawEvent } from "./DrawEvent";
export { User } from "./User";
import { Repository } from "typeorm";
import { Game } from "./Game";
export { Game } from "./Game";
import { LocalCredentials } from "./LocalCredentials";
import { DeviantartAccount } from "./DeviantartAccount";
export { LocalCredentials } from "./LocalCredentials";
/**
 * Interface representing a model constructor function.
 */
export interface IModelClass<T = any> {
  createInitialEntities?: (...args: any[]) => Promise<any>;
  new (...args: any[]): T;
}
export const MODELS: IModelClass[] = [
  User,
  Prize,
  Role,
  DrawEvent,
  Game,
  LocalCredentials,
  DeviantartAccount
];

declare global {
  interface ApplicationContextMembers {
    // see OrmContext for creation of these default repositories
    // if a custom repository is created for any model in this list,
    // remove it from here and declare its more specific type in the
    // implementation file
    roleRepository: Repository<Role>;
    gameRepository: Repository<Game>;
  }
}
