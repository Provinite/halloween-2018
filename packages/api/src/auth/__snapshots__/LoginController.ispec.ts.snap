// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /login validates principal and password 1`] = `
Object {
  "error": "RequestValidationError",
  "errors": Object {
    "password": Object {
      "message": "password must be a non empty string",
      "pass": false,
    },
  },
  "message": "Validation failed for field(s): password",
  "status": 400,
}
`;

exports[`POST /login validates principal and password 2`] = `
Object {
  "error": "RequestValidationError",
  "errors": Object {
    "password": Object {
      "message": "password must be a non empty string",
      "pass": false,
    },
    "principal": Object {
      "message": "principal must be a non empty string",
      "pass": false,
    },
  },
  "message": "Validation failed for field(s): principal, password",
  "status": 400,
}
`;
